src/rpc/daemon_rpc_client.h

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) 2017, The Monero Project
+// 
+// All rights reserved.
+// 
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+// 
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+// 
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+// 
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+// 
+// Parts of this file are originally copyright (c) 2012-2013 The Cryptonote developers
+
+#pragma once
+
+#include <boost/optional/optional.hpp>
+
+#include "rpc/message_data_structs.h"
+#include "net/http_auth.h"
+
+namespace cryptonote
+{
+
+namespace rpc
+{
+
+
+/**
+ * @brief Daemon RPC Client interface
+ *
+ * This class serves to bridge the old RPC with the new ZMQ RPC such that
+ * one can command a wallet to use either to communicate with a daemon.
+ */
+class DaemonRPCClient
+{
+  public:
+
+    enum CLIENT_TYPE
+    {
+      TYPE_OLD,
+      TYPE_ZMQ
+    };
+
+    DaemonRPCClient() { }
+    virtual ~DaemonRPCClient() { }
+
+    /**
+     * @brief Check the connection to the daemon's RPC server
+     *
+     * @param timeout [in] timeout in ms to wait for a reply
+     * @param version [out] the daemon's RPC version
+     * @param error_details [out] details of any RPC error
+     *
+     * @return true if version info obtained, otherwise false
```

Comment:
As far as I looked at other parts of this patch where `boost::optional<std::string>` is used as the return type of a function, the convention seems to be that the return value is set to `boost::none` when the process succeeds, and to an error string when fails. If this convention is meant to be applied here, the comment seems wrong; it should've been something like:
```
@return boost::none if version info obtained, otherwise an error string
```
Also, the function doesn't have an argument named `error_details`. In any case, the inconsistency between the comment and the declaration should be resolved.

Reply:
error_details is the return value I think.

Reply:
I replied to this via email, but it seems github didn't put the replies here as I thought it would.

error_details is indeed the return value.  That changed, but I didn't update the comment accordingly.

FIXME